File: ml_test_vectors/stateless/tanh/implementation.py
from typing import Optional

import numpy as np

from ml_test_vectors.core import Tensor


def forward(inp: Tensor) -> Tensor:
    x = np.array(inp, dtype="float32")
    return np.tanh(x, dtype="float32").tolist()


def backward(inp: Tensor, output: Optional[Tensor]) -> Tensor:
    outp = forward(inp) if output is None else output
    x = np.array(outp, dtype="float32")
    return (1 - x**2).tolist()

File: ml_test_vectors/stateless/tanh/implementation.ts
import type {
	StatelessDerivativeFunction,
	StatelessFunction,
	Tensor,
} from "../../core";
import { isScalar } from "../../utils";

export const forward: StatelessFunction<Tensor, Tensor> = (input) =>
	isScalar(input) ? Math.tanh(input) : input.map(forward);

export const backward: StatelessDerivativeFunction = (
	input,
	output = forward(input),
) => {
	if (isScalar(output)) {
		return 1 - output ** 2;
	}
	if (isScalar(input)) {
		throw new Error("Mismatched input/output");
	}
	return output.map((o, i) => backward(input[i], o));
};
